### PR TITLE
Feat: Introduce support for pre- and post- statements to seeds

### PR DESCRIPTION
**Pull request by izeigerman**

### CODE REVIEW

File: sqlmesh/core/model/definition.py [hunk 1]
```diff
@@ -1289,20 +1327,23 @@ def _create_model(
     return t.cast(Model, model)
 
 
-def _extract_sql_model_select(
+SEED_INSERT_MACRO_CALL = d.parse("@SEED_INSERT()")[0]
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1330]: @tobymao @crericha open to suggestions around naming this macro call

File: sqlmesh/core/model/definition.py [hunk 2]
```diff
@@ -596,7 +581,64 @@ def render_pre_statements(
...
+
+    @property
+    def post_statements(self) -> t.List[exp.Expression]:
+        return self.post_statements_ or []
+
+    @property
+    def macro_definitions(self) -> t.List[d.MacroDef]:
+        """All macro definitions from the list of expressions."""
+        return [
+            s for s in [*self.pre_statements, *self.post_statements] if isinstance(s, d.MacroDef)
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:596]: itertools.chain() or self.pre_statements + self.post_statements

File: sqlmesh/core/model/definition.py [hunk 3]
```diff
@@ -596,7 +581,64 @@ def render_pre_statements(
...
+        ]
+
+    def _render_statements(
+        self,
+        statements: t.Iterable[exp.Expression],
+        **kwargs: t.Any,
+    ) -> t.List[exp.Expression]:
+        rendered = [
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:604]: (
) to make it a gen so two arrays aren't created

File: sqlmesh/core/model/definition.py (hunk 1 above)
Comment (tobymao) [reply sqlmesh/core/model/definition.py:1330]: what about @INSERT_SEED()

**Review by tobymao — APPROVED** [review]

File: sqlmesh/dbt/basemodel.py
Comment (crericha) [inline sqlmesh/dbt/basemodel.py:1]: In BaseModelConfig's docstring, remove "(not supported for seeds)" from pre_hook and post_hook.

File: sqlmesh/dbt/basemodel.py [hunk 4]
```diff
@@ -236,6 +237,16 @@ def sqlmesh_model_kwargs(self, context: DbtContext) -> t.Dict[str, t.Any]:
             "jinja_macros": jinja_macros,
             "path": self.path,
             "hash_raw_query": True,
+            "pre_statements": [
+                exp
+                for hook in self.pre_hook
+                for exp in d.parse(hook.sql, default_dialect=self.model_dialect or context.dialect)
```
Comment (crericha) [inline sqlmesh/dbt/basemodel.py:243]: Include sql_header as the first pre_hook if sql_header is not None. I merged changes today that had this behavior in Model.py, which should be here now instead.

Comment (crericha) [reply sqlmesh/dbt/basemodel.py:243]: Note that the removed code in Model.py used `pre_hooks` var and not `self.pre_hook`

**Comment by crericha** [comment]
Do plan context diffs include the pre/post statements? If so, is there any work that needs to be done there?

File: sqlmesh/dbt/basemodel.py (hunk 4 above)
Comment (izeigerman) [reply sqlmesh/dbt/basemodel.py:243]: good catch, thanks

**Comment by izeigerman** [comment]
> Do plan context diffs include the pre/post statements? If so, is there any work that needs to be done there?

They do

**Pull request merged**